pageserver/src/object_store.rs

```diff
@@ -0,0 +1,64 @@
+use crate::repository::{BufferTag, RelTag};
+use crate::ZTimelineId;
+use anyhow::Result;
+use serde::{Deserialize, Serialize};
+use zenith_utils::lsn::Lsn;
+use std::collections::HashSet;
+use std::iter::Iterator;
+
+#[derive(Debug, Clone)]
+#[derive(Serialize, Deserialize)]
+
+pub struct ObjectKey {
+    pub timeline: ZTimelineId,
+    pub buf_tag: BufferTag,
+}
+
+///
+/// Low-level storage abstraction.
+///
+/// All the data in the repository is stored in a key-value store. This trait
+/// abstracts the details of the key-value store.
+///
+/// A simple key-value store would support just GET and PUT operations with
+/// a key, but the upper layer needs slightly complicated read operations
+///
+/// The most frequently used function is 'block_iter_backwards'. It is used
+/// to look up a page version. It's LSN aware, in that the caller specifies an
+/// LSN, and the function returns all values for that block with the same
+/// or older LSN.
+///
+/// Store contains all the pageserver data of all tenants and don't know
+/// anything about our more high-level concepts like timelines and tenants.
+/// It just provides abstraction of persistent ordered collection.
+pub trait ObjectStore: Send + Sync  {
+
+    ///
+    /// Store a value with given key.
+    ///
+    fn put(&self, key: &ObjectKey, lsn: Lsn, value: &[u8]) -> Result<()>;
+
+    /// Read entry with the exact given key.
+    ///
+    /// This is used for retrieving metadata with special key that doesn't
+    /// correspond to any real relation.
+    fn get(&self, key: &ObjectKey, lsn: Lsn) -> Result<Vec<u8>>;
+
+    /// Iterate through all page versions of one object, in descending LSN order.
+    fn object_versions<'a>(&'a self, key: &ObjectKey, lsn: Lsn) -> Result<Box<dyn ObjectVersionIterator + 'a>>;
+
+    /// Iterate through all keys with given tablespace and database ID, and LSN <= 'lsn'.
+    ///
+    /// This is used to implement 'create database'
+    fn list_rels<'a>(&'a self, timelineid: ZTimelineId, spcnode: u32, dbnode: u32, lsn: Lsn) -> Result<HashSet<RelTag>>;
+
+    // we don't really delete any data in storage we only can throw away some kv
+    // pairs during lsm-like merge. (that may be records with old lsn or whole
+    // database -- callback would receive key and return bool indicating whether
+    // to keep that kv pair or not).
+    //fn set_merge_filter()
+}
+
+pub trait ObjectVersionIterator: Iterator<Item = Result<(Lsn, Vec<u8>)>> {
+
+}
```

Comment:
Add docs, explaining what this is?

Reply:
The idea is to define a type for what `object_versions` returns. Currently, it's:
```
    /// Iterate through all page versions of one object, in descending LSN order.
    fn object_versions<'a>(
        &'a self,
        key: &ObjectKey,
        lsn: Lsn,
    ) -> Result<Box<dyn ObjectVersionIterator + 'a>>;
```
Without this trait, it'd have to be:
```
    /// Iterate through all page versions of one object, in descending LSN order.
    fn object_versions<'a>(
        &'a self,
        key: &ObjectKey,
        lsn: Lsn,
    ) -> Result<Box<dyn Iterator<Item = (Lsn, Vec<u8>)> + 'a>>;
```
Hmm, now that I look at it, it's not too bad. It's actually nice to show the "Lsn" and "Vec<u8>" right here. I'll remove this trait and just use `Iterator<Item = (Lsn, Vec<u8>)>` instead, after all.